src/Microsoft.AspNetCore.Mvc.Core/Controllers/ControllerFeatureProvider.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+
+namespace Microsoft.AspNetCore.Mvc.Controllers
+{
+    /// <summary>
+    /// Discovers controllers from a list of <see cref="ApplicationPart"/> instances.
+    /// </summary>
+    public class ControllerFeatureProvider : IApplicationFeatureProvider<ControllerFeature>
+    {
+        private const string ControllerTypeNameSuffix = "Controller";
+
+        /// <inheritdoc />
+        public void PopulateFeature(
+            IEnumerable<ApplicationPart> parts,
+            ControllerFeature feature)
+        {
+            foreach (var part in parts.OfType<IApplicationPartTypeProvider>())
+            {
+                foreach (var type in part.Types)
+                {
+                    if (IsController(type) && !feature.Controllers.Contains(type))
```

Comment:
Completely agree that the check is necessary, in some manner. I am concerned about a not-unreasonable case where I have an app with 100 controllers in an assembly - this will do ~10,000 checks for `Contains`, right? Would be good to either measure, or come up with an algorithm that intrinsically has minimal cost (not sure exactly what that would be).

And, even though it's a one-time startup cost, we need to keep startup costs as low as possible.